File: practice_in_C++/MergeKSortedLists.cpp
#include "./data_structures/ListNode.h"
#include <memory>
#include <queue>
#include <iostream>
#define null NULL
#define p(x) std::cout << x << std::endl;
#define ll long long
#define ar std::array
using namespace std;

class Solution
{
public:
    //  store all values in each linked list in a min heap then store in
    //  either a new linked list or try to reuse the old ones in the list.
    //  Because the constraint states that the list max size is 10^4 but any
    //  individual linked list can have a max size of 500, it is clear that
    //  it is better to process each individual linked list first for my algorithm.
    ListNode *mergeKLists(vector<ListNode *> &lists)
    {
        if (lists.size() == 0)
            return null;

        // default is a maxheap, so change to min heap with greater<int>
        priority_queue<int, vector<int>, greater<int>> minheap;

        for (auto &lst : lists)
        {
            ListNode *tmp = lst;
            while (tmp != null)
            {
                minheap.emplace(tmp->val);
                tmp = tmp->next;
            }
        }

        // while (!minheap.empty())
        // {
        //     p(minheap.top());
        //     minheap.pop();
        // }

        ListNode *dummy = new ListNode(10001);
        ListNode *tmp = dummy;
        ListNode *last = null;

        for (auto &lst : lists)
        {
            if (lst != null)
            {
                tmp->next = lst;
                tmp = tmp->next;
                break;
            }
        }

        for (auto &lst : lists)
        {
            if (lst == null)
            {
                continue;
            }

            tmp = lst;

            // initially, last will be null
            if (last != null)
            {
                last->next = tmp;
            }

            while (tmp != null)
            {
                tmp->val = minheap.top();
                p(minheap.top())
                minheap.pop();
                last = tmp;
                tmp = tmp->next;
            }
        }

        ListNode *toDelete = dummy;
        dummy = dummy->next;
        delete toDelete;

        return dummy;
    }
};

File: C#/practice_problems/leetcode_questions/MergeKSortedLists.cs
using System.Collections.Generic;
using data_structures;

namespace practice_problems
{
    using ListNode = ListNode<int>;
    class MergeKSortedLists
    {
        public ListNode MergeKLists(ListNode[] lists)
        {
            if (lists.Length == 0)
                return null;

            // default is a maxheap, so change to min heap with greater<int>
            var minheap = new SortedList<int, int>();

            foreach (var lst in lists)
            {
                ListNode aux = lst;
                while (aux != null)
                {
                    minheap[aux.val] = minheap.ContainsKey(aux.val) ? minheap[aux.val] + 1 : 1;
                    aux = aux.next;
                }
            }

            // foreach (var kvp in minheap)
            // {
            //     Console.WriteLine($"{kvp.Key} : {kvp.Value}");
            // }

            ListNode dummy = new ListNode(0);
            ListNode tmp = dummy;
            ListNode last = null;

            foreach (var lst in lists)
            {
                if (lst != null)
                {
                    tmp.next = lst;
                    tmp = tmp.next;
                    break;
                }
            }

            foreach (var lst in lists)
            {
                if (lst == null)
                {
                    continue;
                }

                tmp = lst;

                // initially, last will be null
                if (last != null)
                {
                    last.next = tmp;
                }

                while (tmp != null)
                {
                    tmp.val = minheap.Keys[0];
                    if (--minheap[minheap.Keys[0]] <= 0)
                    {
                        minheap.RemoveAt(0);
                    }

                    last = tmp;
                    tmp = tmp.next;
                }
            }

            dummy = dummy.next;

            return dummy;
        }
    }

}




